File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int curr_max, curr_min, res;
        curr_min = curr_max = res = nums[0];
        
        for (int i = 1; i < nums.size(); ++i) {
            int num = nums[i];
            int temp_max = max(num, max(curr_max * num, curr_min * num));
            curr_min = min(num, min(curr_max * num, curr_min * num));
            curr_max = temp_max;
            res = max(res, curr_max);
        }
        return res;
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums: return 0
        
        curr_max = nums[0]
        curr_min = nums[0]
        res = curr_max
        
        for i in range(1, len(nums)):
            curr = nums[i]
            temp = max(curr_max * curr, curr_min * curr, curr)
            curr_min = min(curr_max * curr, curr_min * curr, curr)
            curr_max = temp
            res = max(curr_max, res)
            
        return res